public/js/test/mochitest/browser_dbg-breaking.js

```diff
@@ -0,0 +1,41 @@
+/* Any copyright is dedicated to the Public Domain.
+ * http://creativecommons.org/publicdomain/zero/1.0/ */
+
+// Tests the breakpoints are hit in various situations.
+
+add_task(function* () {
+  const dbg = yield initDebugger("doc-scripts.html", "scripts.html");
+  const { selectors: { getSelectedSource }, getState } = dbg;
+
+  // Make sure we can set a top-level breakpoint and it will be hit on
+  // reload.
+  yield addBreakpoint(dbg, "scripts.html", 18);
+  reload(dbg);
+  yield waitForPaused(dbg);
+  assertPausedLocation(dbg, "scripts.html", 18);
+  yield resume(dbg);
+
+  const paused = waitForPaused(dbg);
+
+  // Create an eval script that pauses itself.
+  invokeInTab("doEval")
+
+  yield paused;
```

Comment:
It's more a concern of when the `threadClient` gets notifications from the debugger server; `waitForClientEvents("paused")` waits for a future paused event, and if it's already come through it will timeout. We ran into this a lot in the old tests, but I think it was much more problematic because promises were used all throughout the UI so thread events could come through half-way through updating the UI. Our system is much more deterministic.